File: 1_Sum_of_array/sum.py
import time
import numpy as np
from statistics import mean

def timeit(func):
    def wrapper(*args, **kwargs):
        start = time.time()
        result = func(*args, **kwargs)
        end = time.time()
        diff = (end - start) * 1000
        return result, diff
    return wrapper

@timeit
def calculate_sum_python(size: int) -> int:
    return sum(range(size))

@timeit
def calculate_sum_numpy(size: int) -> int:
    return np.sum(np.arange(size))

if __name__ == "__main__":
    SIZE = 10_000_000
    ITERATIONS = 10
    
    python_times = []
    numpy_times = []
    
    # Run benchmarks
    for _ in range(ITERATIONS):
        _, python_time = calculate_sum_python(SIZE)
        python_times.append(python_time)
        
        _, numpy_time = calculate_sum_numpy(SIZE)
        numpy_times.append(numpy_time)
    
    # Calculate statistics
    print("Python Implementation Results (ms):")
    print(f"Min: {min(python_times):.2f}, Max: {max(python_times):.2f}, Avg: {mean(python_times):.2f}\n")
    
    print("NumPy Implementation Results (ms):")
    print(f"Min: {min(numpy_times):.2f}, Max: {max(numpy_times):.2f}, Avg: {mean(numpy_times):.2f}")

# Python Implementation Results (ms):
# Min: 141.05, Max: 189.50, Avg: 158.68

# NumPy Implementation Results (ms):
# Min: 17.80, Max: 61.88, Avg: 23.71

File: 1_Sum_of_array/sum.cpp
#include <iostream>
#include <chrono>
#include <numeric>
#include <vector>
#include <iomanip>
#include <algorithm>

// Decorator-like timing function template that returns duration
template<typename F, typename... Args>
auto measureTime(F func, Args... args) {
    auto start = std::chrono::high_resolution_clock::now();
    auto result = func(args...);
    auto end = std::chrono::high_resolution_clock::now();
    
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    return std::make_pair(result, duration.count());
}

// Standard C++ sum implementation
long long calculateSumCpp(int size) {
    long long sum = 0;
    for (int i = 0; i < size; ++i) {
        sum += i;
    }
    return sum;
}

// STL algorithm implementation using std::accumulate
long long calculateSumSTL(int size) {
    std::vector<int> numbers(size);
    std::iota(numbers.begin(), numbers.end(), 0);
    return std::accumulate(numbers.begin(), numbers.end(), 0LL);
}

int main() {
    const int SIZE = 10000000;
    const int ITERATIONS = 10;
    
    std::vector<long long> cpp_times;
    std::vector<long long> stl_times;
    
    // Run benchmarks
    for (int i = 0; i < ITERATIONS; ++i) {
        auto [result_cpp, time_cpp] = measureTime(calculateSumCpp, SIZE);
        cpp_times.push_back(time_cpp);
        
        auto [result_stl, time_stl] = measureTime(calculateSumSTL, SIZE);
        stl_times.push_back(time_stl);
    }
    
    // Calculate statistics for C++ implementation
    auto cpp_min = *std::min_element(cpp_times.begin(), cpp_times.end());
    auto cpp_max = *std::max_element(cpp_times.begin(), cpp_times.end());
    auto cpp_avg = std::accumulate(cpp_times.begin(), cpp_times.end(), 0LL) / ITERATIONS;
    
    // Calculate statistics for STL implementation
    auto stl_min = *std::min_element(stl_times.begin(), stl_times.end());
    auto stl_max = *std::max_element(stl_times.begin(), stl_times.end());
    auto stl_avg = std::accumulate(stl_times.begin(), stl_times.end(), 0LL) / ITERATIONS;
    
    // Print results
    std::cout << std::fixed << std::setprecision(2);
    std::cout << "C++ Implementation Results (ms):\n";
    std::cout << "Min: " << cpp_min << ", Max: " << cpp_max << ", Avg: " << cpp_avg << "\n\n";
    
    std::cout << "STL Implementation Results (ms):\n";
    std::cout << "Min: " << stl_min << ", Max: " << stl_max << ", Avg: " << stl_avg << "\n";
    
    return 0;
}

// C++ Implementation Results (ms):
// Min: 24, Max: 26, Avg: 25

// STL Implementation Results (ms):
// Min: 219, Max: 280, Avg: 235